File: VB/PDF Viewer/Form1.vb
Imports DevExpress.Pdf
Imports DevExpress.XtraBars.Ribbon
Imports DevExpress.XtraPdfViewer
Imports DevExpress.XtraPdfViewer.Commands
Imports DevExpress.XtraPdfViewer.Localization
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Windows.Forms

Namespace PDF_Viewer

    Public Partial Class Form1
        Inherits RibbonForm

        Private Class CustomSaveAsCommand
            Inherits PdfSaveAsFileCommand

            Private ReadOnly rectangles As IList(Of GraphicsRect)

            Public Sub New(ByVal control As PdfViewer, ByVal rectangles As IList(Of GraphicsRect))
                MyBase.New(control)
                Me.rectangles = rectangles
            End Sub

            Public Overrides Sub Execute()
                If rectangles Is Nothing OrElse rectangles.Count = 0 Then
                    MyBase.Execute()
                    Return
                End If

                Using saveFileDialog As SaveFileDialog = New SaveFileDialog()
                    saveFileDialog.OverwritePrompt = True
                    saveFileDialog.Filter = XtraPdfViewerLocalizer.GetString(XtraPdfViewerStringId.PDFFileFilter)
                    If saveFileDialog.ShowDialog() <> DialogResult.OK Then Return
                    Using processor As PdfDocumentProcessor = New PdfDocumentProcessor()
                        processor.LoadDocument(Control.DocumentFilePath)
                        For Each rect In rectangles
                            Using graph As PdfGraphics = processor.CreateGraphics()
                                Dim page As PdfPage = processor.Document.Pages(rect.PageIndex)
                                Dim pageCropBox As PdfRectangle = page.CropBox
                                Dim p1 As PdfPoint = New PdfPoint(rect.Point1.X, pageCropBox.Height - rect.Point1.Y)
                                Dim p2 As PdfPoint = New PdfPoint(rect.Point2.X, pageCropBox.Height - rect.Point2.Y)
                                Dim bounds As RectangleF = RectangleF.FromLTRB(CSng(Math.Min(p1.X, p2.X)), CSng(Math.Min(p1.Y, p2.Y)), CSng(Math.Max(p1.X, p2.X)), CSng(Math.Max(p1.Y, p2.Y)))
                                graph.DrawRectangle(New Pen(Color.Red), bounds)
                                graph.AddToPageForeground(page, 72, 72)
                            End Using
                        Next

                        processor.SaveDocument(saveFileDialog.FileName)
                    End Using
                End Using
            End Sub
        End Class

        Private Class CustomCommandService
            Implements IPdfViewerCommandFactoryService

            Private ReadOnly viewer As PdfViewer

            Private ReadOnly service As IPdfViewerCommandFactoryService

            Public Property Rectangles As IList(Of GraphicsRect)

            Public Sub New(ByVal service As IPdfViewerCommandFactoryService, ByVal viewer As PdfViewer)
                Me.viewer = viewer
                Me.service = service
            End Sub

            Public Function CreateCommand(ByVal commandId As PdfViewerCommandId) As PdfViewerCommand Implements IPdfViewerCommandFactoryService.CreateCommand
                If commandId = PdfViewerCommandId.SaveAsFile Then Return New CustomSaveAsCommand(viewer, Rectangles)
                Return service.CreateCommand(commandId)
            End Function
        End Class

        Private Class GraphicsRect

            Public Sub New(ByVal pageIndex As Integer, ByVal point1 As PdfPoint, ByVal point2 As PdfPoint)
                Me.PageIndex = pageIndex
                Me.Point1 = point1
                Me.Point2 = point2
            End Sub

            Public ReadOnly Property PageIndex As Integer

            Public ReadOnly Property Point1 As PdfPoint

            Public ReadOnly Property Point2 As PdfPoint

            Public ReadOnly Property IsEmpty As Boolean
                Get
                    Return Point1 = Point2
                End Get
            End Property
        End Class

        Private imageRectangleList As List(Of GraphicsRect) = New List(Of GraphicsRect)()

        Private currentImageRect As GraphicsRect

        Private ActivateDrawing As Boolean = False

        Private commandService As CustomCommandService

        Public Sub New()
            InitializeComponent()
            pdfViewer.LoadDocument("Demo.pdf")
            AddHandler pdfViewer.MouseDown, AddressOf pdfViewer1_MouseDown
            AddHandler pdfViewer.MouseUp, AddressOf pdfViewer1_MouseUp
            AddHandler pdfViewer.MouseMove, AddressOf PdfViewer_MouseMove
            AddHandler pdfViewer.Paint, AddressOf PdfViewer_Paint
            pdfViewer.CursorMode = PdfCursorMode.Custom
            AddHandler pdfViewer.MouseWheel, AddressOf PdfViewer_MouseWheel
            Dim service = pdfViewer.GetService(Of IPdfViewerCommandFactoryService)()
            pdfViewer.RemoveService(GetType(IPdfViewerCommandFactoryService))
            commandService = New CustomCommandService(service, pdfViewer)
            commandService.Rectangles = imageRectangleList
            pdfViewer.AddService(GetType(IPdfViewerCommandFactoryService), commandService)
        End Sub

        Private Sub PdfViewer_MouseWheel(ByVal sender As Object, ByVal e As MouseEventArgs)
            If ModifierKeys = Keys.Control Then
                pdfViewer.ZoomFactor += e.Delta \ 100
            Else
                pdfViewer.ScrollVertical(-e.Delta)
            End If
        End Sub

        Private Sub UpdateCurrentRect(ByVal location As Point)
            If currentImageRect IsNot Nothing Then
                Dim documentPosition = pdfViewer.GetDocumentPosition(location, True)
                If currentImageRect.PageIndex = documentPosition.PageNumber - 1 Then currentImageRect = New GraphicsRect(currentImageRect.PageIndex, currentImageRect.Point1, documentPosition.Point)
            End If
        End Sub

        Private Sub PdfViewer_MouseMove(ByVal sender As Object, ByVal e As MouseEventArgs)
            If currentImageRect IsNot Nothing Then
                UpdateCurrentRect(e.Location)
                pdfViewer.Invalidate()
            End If
        End Sub

        Private Sub pdfViewer1_MouseUp(ByVal sender As Object, ByVal e As MouseEventArgs)
            UpdateCurrentRect(e.Location)
            If currentImageRect IsNot Nothing Then
                If Not currentImageRect.IsEmpty AndAlso ActivateDrawing Then imageRectangleList.Add(currentImageRect)
                currentImageRect = Nothing
            End If
        End Sub

        Private Sub pdfViewer1_MouseDown(ByVal sender As Object, ByVal e As MouseEventArgs)
            Dim position = pdfViewer.GetDocumentPosition(e.Location, True)
            currentImageRect = New GraphicsRect(position.PageNumber - 1, position.Point, position.Point)
        End Sub

        Private Sub DrawImageRectangle(ByVal graphics As Graphics, ByVal rect As GraphicsRect)
            Dim start As PointF = pdfViewer.GetClientPoint(New PdfDocumentPosition(rect.PageIndex + 1, rect.Point1))
            Dim [end] As PointF = pdfViewer.GetClientPoint(New PdfDocumentPosition(rect.PageIndex + 1, rect.Point2))
            Dim r = Rectangle.FromLTRB(CInt(Math.Min(start.X, [end].X)), CInt(Math.Min(start.Y, [end].Y)), CInt(Math.Max(start.X, [end].X)), CInt(Math.Max(start.Y, [end].Y)))
            graphics.DrawRectangle(New Pen(Color.Red), r)
        End Sub

        Private Sub PdfViewer_Paint(ByVal sender As Object, ByVal e As PaintEventArgs)
            If ActivateDrawing Then
                For Each r In imageRectangleList
                    DrawImageRectangle(e.Graphics, r)
                Next

                If currentImageRect IsNot Nothing Then DrawImageRectangle(e.Graphics, currentImageRect)
            End If
        End Sub

        Private Sub SaveDrawingAndReload()
            Dim fileName As String = pdfViewer.DocumentFilePath
            pdfViewer.CloseDocument()
            Using processor As PdfDocumentProcessor = New PdfDocumentProcessor()
                processor.LoadDocument(fileName)
                For Each rect In imageRectangleList
                    Using graph As PdfGraphics = processor.CreateGraphics()
                        Dim page As PdfPage = processor.Document.Pages(rect.PageIndex)
                        Dim pageCropBox As PdfRectangle = page.CropBox
                        Dim p1 As PdfPoint = New PdfPoint(rect.Point1.X, pageCropBox.Height - rect.Point1.Y)
                        Dim p2 As PdfPoint = New PdfPoint(rect.Point2.X, pageCropBox.Height - rect.Point2.Y)
                        Dim bounds As RectangleF = RectangleF.FromLTRB(CSng(Math.Min(p1.X, p2.X)), CSng(Math.Min(p1.Y, p2.Y)), CSng(Math.Max(p1.X, p2.X)), CSng(Math.Max(p1.Y, p2.Y)))
                        graph.DrawRectangle(New Pen(Color.Red), bounds)
                        graph.AddToPageForeground(page, 72, 72)
                    End Using
                Next

                processor.SaveDocument(fileName)
            End Using

            imageRectangleList.Clear()
            pdfViewer.LoadDocument(fileName)
        End Sub

        Private Sub barButtonItem2_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs)
            ActivateDrawing = Not ActivateDrawing
            pdfViewer.Invalidate()
        End Sub

        Private Sub barButtonItem3_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs)
            SaveDrawingAndReload()
        End Sub
    End Class
End Namespace

File: CS/PDF Viewer/Form1.cs
using DevExpress.Pdf;
using DevExpress.XtraBars.Ribbon;
using DevExpress.XtraPdfViewer;
using DevExpress.XtraPdfViewer.Commands;
using DevExpress.XtraPdfViewer.Localization;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;

namespace PDF_Viewer
{
    public partial class Form1 : RibbonForm
    {
        #region SaveAsCommand
        class CustomSaveAsCommand : PdfSaveAsFileCommand
        {
            readonly IList<GraphicsCoordinates> rectangles;

            public CustomSaveAsCommand(PdfViewer control, IList<GraphicsCoordinates> rectangles) : base(control)
            {
                this.rectangles = rectangles;
            }
            public override void Execute()
            {

                if (rectangles == null || rectangles.Count == 0)
                {
                    base.Execute();
                    return;
                }
                using (SaveFileDialog saveFileDialog = new SaveFileDialog())
                {
                    saveFileDialog.OverwritePrompt = true;
                    saveFileDialog.Filter = XtraPdfViewerLocalizer.GetString(XtraPdfViewerStringId.PDFFileFilter);
                    if (saveFileDialog.ShowDialog() != DialogResult.OK)
                        return;
                    using (PdfDocumentProcessor processor = new PdfDocumentProcessor())
                    {
                        processor.LoadDocument(Control.DocumentFilePath);
                        foreach (var rect in rectangles)
                        {
                            using (PdfGraphics graph = processor.CreateGraphics())
                            {
                                PdfPage page = processor.Document.Pages[rect.PageIndex];
                                PdfRectangle pageCropBox = page.CropBox;
                                PdfPoint p1 = new PdfPoint(rect.Point1.X, pageCropBox.Height - rect.Point1.Y);
                                PdfPoint p2 = new PdfPoint(rect.Point2.X, pageCropBox.Height - rect.Point2.Y);
                                RectangleF bounds = RectangleF.FromLTRB(
                                    (float)Math.Min(p1.X, p2.X), (float)Math.Min(p1.Y, p2.Y),
                                    (float)Math.Max(p1.X, p2.X), (float)Math.Max(p1.Y, p2.Y));
                                graph.DrawRectangle(new Pen(Color.Red), bounds);
                                graph.AddToPageForeground(page, 72, 72);
                            }
                        }
                        processor.SaveDocument(saveFileDialog.FileName);
                    }
                }
            }
        }


        class CustomCommandService : IPdfViewerCommandFactoryService
        {
            readonly PdfViewer viewer;
            readonly IPdfViewerCommandFactoryService service;

            public IList<GraphicsCoordinates> Rectangles { get; set; }

            public CustomCommandService(IPdfViewerCommandFactoryService service, PdfViewer viewer)
            {
                this.viewer = viewer;
                this.service = service;
            }
            public PdfViewerCommand CreateCommand(PdfViewerCommandId commandId)
            {
                if (commandId == PdfViewerCommandId.SaveAsFile)
                    return new CustomSaveAsCommand(viewer, Rectangles);
                return service.CreateCommand(commandId);
            }
        }
        #endregion SaveAsCommand

        #region GraphicsCoordinates

        // This class is used to save
        // and restore the selection area coordinates
        class GraphicsCoordinates
        {
            public GraphicsCoordinates(int pageIndex, PdfPoint point1, PdfPoint point2) {
                PageIndex = pageIndex;
                Point1 = point1;
                Point2 = point2;
            }

            public int PageIndex { get; }
            public PdfPoint Point1 { get; }
            public PdfPoint Point2 { get; }
            public bool IsEmpty => Point1 == Point2;
        }

        List<GraphicsCoordinates> rectangleCoordinateList = new List<GraphicsCoordinates>();
        GraphicsCoordinates currentCoordinates;

        // This variable indicates whether the Drawing button
        // is activated
        bool ActivateDrawing = false;
        #endregion GraphicsCoordinates

        CustomCommandService commandService;

        public Form1() {
            InitializeComponent();
            pdfViewer.LoadDocument("Demo.pdf");
            pdfViewer.MouseDown += pdfViewer1_MouseDown;
            pdfViewer.MouseUp += pdfViewer1_MouseUp;
            pdfViewer.MouseMove += PdfViewer_MouseMove;
            pdfViewer.Paint += PdfViewer_Paint;
            pdfViewer.CursorMode = PdfCursorMode.Custom;
            pdfViewer.MouseWheel += PdfViewer_MouseWheel;

            var service = pdfViewer.GetService<IPdfViewerCommandFactoryService>();
            pdfViewer.RemoveService(typeof(IPdfViewerCommandFactoryService));
            commandService = new CustomCommandService(service, pdfViewer);
            commandService.Rectangles = rectangleCoordinateList;
            pdfViewer.AddService(typeof(IPdfViewerCommandFactoryService), commandService);
        }

        void PdfViewer_MouseWheel(object sender, MouseEventArgs e) {
            if (ModifierKeys == Keys.Control)
                pdfViewer.ZoomFactor += e.Delta / 100;
            else
                pdfViewer.ScrollVertical(-e.Delta);
        }

        #region MouseEvents
        void PdfViewer_MouseMove(object sender, MouseEventArgs e) {
            if (currentCoordinates != null) {
                UpdateCurrentRect(e.Location);
                pdfViewer.Invalidate();
            }
        }
        void pdfViewer1_MouseUp(object sender, MouseEventArgs e) {

            // Convert the retrieved coordinates 
            // to the page coordinates
            UpdateCurrentRect(e.Location);
            if (currentCoordinates != null) {
                if (!currentCoordinates.IsEmpty && ActivateDrawing)
                    // Add coordinates to the list
                    rectangleCoordinateList.Add(currentCoordinates);
                currentCoordinates = null;
            }
        }
        void pdfViewer1_MouseDown(object sender, MouseEventArgs e) {
            var position = pdfViewer.GetDocumentPosition(e.Location, true);
            currentCoordinates = new GraphicsCoordinates(position.PageNumber - 1, position.Point, position.Point);
        }

        void UpdateCurrentRect(Point location)
        {
            if (rectangleCoordinateList != null)
            {
                var documentPosition = pdfViewer.GetDocumentPosition(location, true);
                if (currentCoordinates.PageIndex == documentPosition.PageNumber - 1)
                    
                    currentCoordinates = new GraphicsCoordinates(currentCoordinates.PageIndex, currentCoordinates.Point1, documentPosition.Point);
            }
        }
        #endregion MouseEvents

        #region ActivateDrawing
        private void activateDrawingButton_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
        {
            // Changle the activation indicator
            ActivateDrawing = !ActivateDrawing;
            pdfViewer.Invalidate();
        }

        void PdfViewer_Paint(object sender, PaintEventArgs e) {
            if (ActivateDrawing)
            {
                foreach (var r in rectangleCoordinateList)
                    DrawImageRectangle(e.Graphics, r);
                if (currentCoordinates != null)
                    DrawImageRectangle(e.Graphics, currentCoordinates);
            }
        }

        void DrawImageRectangle(Graphics graphics, GraphicsCoordinates rect)
        {            
            PointF start = pdfViewer.GetClientPoint(new PdfDocumentPosition(rect.PageIndex + 1, rect.Point1));
            PointF end = pdfViewer.GetClientPoint(new PdfDocumentPosition(rect.PageIndex + 1, rect.Point2));
            // Create a rectangle where graphics should be drawn
            var r = Rectangle.FromLTRB((int)Math.Min(start.X, end.X), (int)Math.Min(start.Y, end.Y), (int)Math.Max(start.X, end.X), (int)Math.Max(start.Y, end.Y));
            
            // Draw a rectangle in the created area
            graphics.DrawRectangle(new Pen(Color.Red), r);
        }
        #endregion ActivateDrawing

        #region SaveGrahpics

        private void saveGraphicsButton_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
        {
            SaveDrawingAndReload();
        }

        private void SaveDrawingAndReload()
        {
            string fileName = pdfViewer.DocumentFilePath;
            pdfViewer.CloseDocument();
            using (PdfDocumentProcessor processor = new PdfDocumentProcessor())
            {
                // Load a document to the PdfDocumentProcessor instance
                processor.LoadDocument(fileName);
                foreach (var rect in rectangleCoordinateList)
                {
                    // Create a PdfGraphics object
                    using (PdfGraphics graph = processor.CreateGraphics())
                    {
                        PdfPage page = processor.Document.Pages[rect.PageIndex];
                        PdfRectangle pageCropBox = page.CropBox;
                        PdfPoint p1 = new PdfPoint(rect.Point1.X, pageCropBox.Height - rect.Point1.Y);
                        PdfPoint p2 = new PdfPoint(rect.Point2.X, pageCropBox.Height - rect.Point2.Y);

                        // Create a rectangle where graphics should be drawn
                        RectangleF bounds = RectangleF.FromLTRB(
                            (float)Math.Min(p1.X, p2.X), (float)Math.Min(p1.Y, p2.Y),
                            (float)Math.Max(p1.X, p2.X), (float)Math.Max(p1.Y, p2.Y));
                        // Draw a rectangle in the created area
                        graph.DrawRectangle(new Pen(Color.Red), bounds);
                        
                        // Draw graphics content into a file
                        graph.AddToPageForeground(page, 72, 72);
                    }
                }
                // Save the document
                processor.SaveDocument(fileName);
            }
            rectangleCoordinateList.Clear();

            // Open the document in the PDF Viewer
            pdfViewer.LoadDocument(fileName);
        }
        #endregion SaveGraphics
    }
}
